### PR TITLE
Better detect equivalence in our strong name provider

### PR DESCRIPTION
**Pull request by ToddGrun**
Addresses https://github.com/dotnet/roslyn/issues/74058

LiveUnitTest team reported a fairly recent regression where Roslyn started reporting Project changed workspace events during file addition / removal. This causes the LUT window to unnecessarilly rerun tests.

This regressed as part of a change I made back in March around passing in a temp path to the compiler's strong name provider, as otherwise an EmitStream call fails, breaking intellisense in projects using assembly signing. (Change here: https://github.com/dotnet/roslyn/pull/72587)

The regression is due to me changing several callers to creating a strong name provider based on the path, instead of them all using the default instance. With the previous behavior, equality checks on these newly created StrongNameFileSystem and DesktopStrongNameProvider instances would return false, even though the temp path in the file systems were the same.

It's a bit of a windy path to how this relates to the LUT window. The Roslyn ProjectSystemProject.ChangeProjectProperty call is the crux of where the issue was being exposed. UpdateProjectOptions_NoLock was calling it with a new DesktopStrongNameProvider created with the temp path as outlined above. ChangeProjectProperty checks the old value of the strong name provider against the newValue, and if they are the same, it doesn't add anything to the batched events it will later send out. However, in this case, the equality check failed, so we started sending out a project changed notification, breaking the LUT behavior.

The fix here is to simply implement Equals in StrongNameFileSystem, and have DesktopStrongNameProvider use that instead of just checking for reference equality.

### CODE REVIEW

**Comment by ToddGrun** [comment]
Adding @333fred and @jaredpar as they have a little bit of context with the PR in which I broke this: https://github.com/dotnet/roslyn/pull/72587

File: src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs [hunk 1]
```diff
@@ -31,6 +31,46 @@ public void RespectNullTempPath()
...
+
+        [Fact]
+        public void InequalityUsingKeyFileSearchPaths()
+        {
+            var tempDir = Temp.CreateDirectory();
+            var provider1 = new DesktopStrongNameProvider(keyFileSearchPaths: [@"c:\test"]);
+            var provider2 = new DesktopStrongNameProvider(keyFileSearchPaths: [@"c:\test2"]);
+
+            Assert.NotEqual(provider1, provider2);
+        }
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs:72]: Do you want to add one more test for case sensitivity also counting as unequal?

**Review by jasonmalinowski — COMMENTED** [review]
Should we also add a test at the workspaces layer so we don't regress something like this again?

File: src/Compilers/Core/Portable/StrongName/StrongNameFileSystem.cs [hunk 2]
```diff
@@ -41,5 +41,26 @@ internal virtual bool FileExists(string? fullPath)
         }
 
         internal string? GetSigningTempPath() => _signingTempPath;
+
+        public override int GetHashCode()
+        {
+            return _signingTempPath?.GetHashCode() ?? 0;
+        }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/StrongName/StrongNameFileSystem.cs:48]:
```suggestion
        => StringComparer.Ordinal.GetHashCode(_signingTempPath);
```

File: src/Compilers/Core/Portable/StrongName/StrongNameFileSystem.cs [hunk 3]
```diff
@@ -41,5 +41,26 @@ internal virtual bool FileExists(string? fullPath)
...
+                return true;
+            }
+
+            if (GetType() != obj!.GetType())
+            {
+                return false;
+            }
+
+            var other = (StrongNameFileSystem)obj;
+            return string.Equals(_signingTempPath, other._signingTempPath, StringComparison.Ordinal);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/StrongName/StrongNameFileSystem.cs:63]:
```suggestion
        public override bool Equals(object? obj)
            => Equals(obj as StrongNameFileSystem);

        private bool Equals(StringNameFileSystem? other)
            => this == other || StringComparer.Ordinal.Equals(_signingTempPath, other?._signingTempPath);
```

File: src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs:72]: Done!

**Comment by ToddGrun** [comment]
> Should we also add a test at the workspaces layer so we don't regress something like this again?

Is there an existing test class that this would fit well into? I took a quick look earlier, and didn't see any somewhat easy way to test this change at that level.

**Comment by ToddGrun** [comment]
@333fred and @jaredpar for the compiler reviews

**Review by jaredpar — APPROVED** [review]

**Review by 333fred — APPROVED** [review]

**Pull request merged**